### PR TITLE
Fixed up a new Structured Output Mode 

### PR DESCRIPTION
**Pull request by ivanleomk**
This adds support for OpenAI's new structured output mode with instructor using the same conventional api that we're using otherwise.

----

| :rocket: | This description was created by [Ellipsis](https://www.ellipsis.dev) for commit dd2f6c73be57dca9d757719f7923fb16c7a2039c  | 
|--------|--------|

### Summary:
Introduced structured output mode for OpenAI API with updates for compatibility across client, response processing, tests, and dependencies.

**Key points**:
- Added support for OpenAI's structured output mode.
- Updated `Mode.STRUCTURED_OUTPUTS` in `instructor/mode.py`.
- Enhanced `instructor/client.py` for structured output handling.
- Modified `from_openai` function to use `client.beta.chat.completions.parse`.
- Adjusted type hints and overloads in `instructor/client.py`.
- Updated `instructor/function_calls.py` for structured output parsing.
- Adjusted `instructor/process_response.py` for structured outputs.
- Updated `instructor/retry.py` for structured output mode and validation.
- Changed test client initialization in `instructor.patch`.
- Removed `STRUCTURED_OUTPUTS` from `tests/llm/test_openai/util.py`.
- Updated dependencies in `.github/workflows/pyright.yml` and `pyproject.toml` to version `1.4.0`.
- Fixed `dump_message` in `instructor/utils.py` for JSON handling.

----
Generated with :heart: by [ellipsis.dev](https://www.ellipsis.dev)

### CODE REVIEW

File: instructor/function_calls.py [hunk 1]
```diff
@@ -271,8 +271,21 @@ def from_response(
         if mode in {Mode.JSON, Mode.JSON_SCHEMA, Mode.MD_JSON}:
             return cls.parse_json(completion, validation_context, strict)
 
+        if mode in {Mode.STRUCTURED_OUTPUTS}:
```
[the remark this answers is not in the document]
Comment (ivanleomk) [reply instructor/function_calls.py:274]: message.refusal is a property defined on the sdk according to OpenAI's provided examples

[the remark this answers is not in the document]
Comment (ivanleomk) [reply instructor/function_calls.py:274]:
```
from pydantic import BaseModel

from openai import OpenAI


class Step(BaseModel):
    explanation: str
    output: str


class MathResponse(BaseModel):
    steps: list[Step]
    final_answer: str


client = OpenAI()

completion = client.beta.chat.completions.parse(
    model="gpt-4o-2024-08-06",
    messages=[
        {"role": "system", "content": "You are a helpful math tutor."},
        {"role": "user", "content": "solve 8x + 31 = 2"},
    ],
    response_format=MathResponse,
)

message = completion.choices[0].message
if message.parsed:
    print(message.parsed.steps)
    print(message.parsed.final_answer)
else:
    print(message.refusal)
```

File: instructor/function_calls.py [hunk 2]
```diff
@@ -271,8 +271,21 @@ def from_response(
...
 
+    @classmethod
+    def parse_structured_output(
+        cls: type[BaseModel],
+        completion: ChatCompletion,
+    ):
+        message = completion.choices[0].message
+        if message.refusal:
+            raise ValueError(f"Refusal of {message.refusal.reason}")
+        return message.parsed
```
Comment (ivanleomk) [inline instructor/function_calls.py:287]: Sample API from OpenAI

```
from pydantic import BaseModel

from openai import OpenAI


class Step(BaseModel):
    explanation: str
    output: str


class MathResponse(BaseModel):
    steps: list[Step]
    final_answer: str


client = OpenAI()

completion = client.beta.chat.completions.parse(
    model="gpt-4o-2024-08-06",
    messages=[
        {"role": "system", "content": "You are a helpful math tutor."},
        {"role": "user", "content": "solve 8x + 31 = 2"},
    ],
    response_format=MathResponse,
)

message = completion.choices[0].message
if message.parsed:
    print(message.parsed.steps)
    print(message.parsed.final_answer)
else:
    print(message.refusal)
```

File: instructor/retry.py [hunk 3]
```diff
@@ -144,6 +144,12 @@ def retry_sync(
 
         total_usage = AnthropicUsage(input_tokens=0, output_tokens=0)
 
+    if mode in {Mode.STRUCTURED_OUTPUTS} and validation_context:
```
[the remark this answers is not in the document]
Comment (ivanleomk) [reply instructor/retry.py:147]: Users get a warning instead of a value error - slightly better

File: instructor/client.py [hunk 4]
```diff
@@ -425,12 +425,18 @@ def from_openai(
...
             instructor.Mode.MD_JSON,
+            instructor.Mode.STRUCTURED_OUTPUTS,
         }
 
     if isinstance(client, openai.OpenAI):
         return Instructor(
             client=client,
-            create=instructor.patch(create=client.chat.completions.create, mode=mode),
+            create=instructor.patch(
+                create=client.chat.completions.create
```
Comment (jxnl) [inline instructor/client.py:435]: move this to the line above `return`

File: instructor/function_calls.py [hunk 5]
```diff
@@ -271,8 +271,21 @@ def from_response(
...
+
         raise ValueError(f"Invalid patch mode: {mode}")
 
+    @classmethod
+    def parse_structured_output(
+        cls: type[BaseModel],
+        completion: ChatCompletion,
+    ):
+        message = completion.choices[0].message
+        if message.refusal:
```
Comment (jxnl) [inline instructor/function_calls.py:285]: check hasattr?

File: pyproject.toml [hunk 6]
```diff
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "instructor"
-version = "1.3.7"
+version = "1.3.8"
```
Comment (jxnl) [inline pyproject.toml:3]:
```suggestion
version = "1.4.0"
```

**Review by jxnl — APPROVED** [review]
let me know when all the tests are passing and lets make sure to update the docs or even come out with a small blogpost but 

1) share benchmarks
2) limitations
3) why instructor still has more features

**Comment by zsiegel92** [comment]
> why instructor still has more features

To help grease the wheels for whoever is working on this

* arbitrary Python-executed `@model_validator` code ofc.
* retries when these validators fail, passing error messages back to LLM
* default parameters and optional fields are fine with Instructor

**Comment by ivanleomk** [comment]
Sure thing. I kind of wanna let their API stabilise a bit more too before pushing out a new release, there's a lot of things they don't support (Eg. Streaming ) and I'd like to document some of these individual modes before we push it out,.

**Comment by KennyVaneetvelde** [comment]
They also don't support certain things inside the schema definitions such as "min_length", "default", ... I don't have an exhaustive list, but I believe instructor should filter these out and treat those in the old way through retrying

**Comment by guvkay20** [comment]
We would greatly appreciate expedient merger of this PR. If it is not merged in the next few days it is likely we will need to abandon instructor.

**Review by jxnl — APPROVED** [review]

**Comment by jxnl** [comment]
aside from documentation what else needs to be done?

**Comment by KennyVaneetvelde** [comment]
@jxnl Check my comment, unless you suggest doing that in a separate PR

**Comment by jxnl** [comment]
> We would greatly appreciate expedient merger of this PR. If it is not merged in the next few days it is likely we will need to abandon instructor.

That's fine. Can't say no to one less dependency. 

Just be warned that many things in Pydantic are not supported and it's slower.

**Comment by onejgordon** [comment]
We're very interested in this feature as well as a way to streamline conversion to the structured API as soon as possible, and the other instructor functionality feels more than valuable enough to justify its continued use. But, could anyone shed light on why "it's slower"?

**Comment by ivanleomk** [comment]
> We're very interested in this feature as well as a way to streamline conversion to the structured API as soon as possible, and the other instructor functionality feels more than valuable enough to justify its continued use. But, could anyone shed light on why "it's slower"?

From my understanding there are two main things that make it slower

1. It's using constrained sampling behind the scenes - this means that they need to manually parse your json schema and convert that to a grammar protocol. This can take up to 10s for more complicated issues
2. it might be faster for subsequent requests but these structured outputs are no longer under the zero data retention policy, they do cache some of your data as a result to make it faster.

General 2 cents is that most models work decently well with the normal. tool calling, might as well use that and hook into retries

**Comment by ivanleomk** [comment]
Implemented a better version in #938 which adds support for streaming, iterables and more

**Pull request closed without merging**